### DISCUSSION TITLE
Collection expressions - spread operator Symbol question

### DISCUSSION THREAD

**Discussion by RNabla** [dotnet/csharplang#7686] [Q&A] [answered]
I've learned about collection expressions feature from - https://devblogs.microsoft.com/dotnet/announcing-csharp-12/

While it seems nice addition to language I feel like syntax for spread operator is weird. 

"Looking at the surrounding ecosystem, we also find examples everywhere of list creation being more convenient and pleasant to use. TypeScript, Dart, Swift, Elm, Python, and more opt for a succinct syntax for this purpose, with widespread usage, and to great effect. Cursory investigations have revealed no substantive problems arising in those ecosystems with having these literals built in."
Taken from https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#motivation

While the concept is familiar, syntax seems unfamiliar. The first question that came to my mind - why not 3 dots like it's in JavaScript, Dart and even Java.

I feel like using something familiar (for developers coming from other languages) would be a better design choice. 
Is there reason for choosing .. instead of any other symbol? I've tried searching for such information, however it seems quite futile due to operator not being letters

**Comment by CyrusNajmabadi** [comment, marked as answer]
> While the concept is familiar, syntax seems unfamiliar. The first question that came to my mind - why not 3 dots like it's in JavaScript, Dart and even Java.

This is covered in the design doc :)

Specifically, we already use `..` to represent a set of elements when deconstructing.  Like so:

```c#
if (x is [var start, .. var middle, var end])
```

We also want construction and deconstruction to be mirrored.  So the form that allows that is:

```c#
M([start, .. middle, end]);
```

> syntax seems unfamiliar.

Syntax familiarity comes *very* quickly to developers.  

> I feel like using something familiar (for developers coming from other languages) would be a better design choice.

We do consider that.  And *all other things being equal* we might have landed there.  But in this case, all other things were not equal.  Strong parity in our own language def takes precedent.

**Reply by RNabla** [reply]
> Strong parity in our own language def takes precedent.
> Specifically, we already use .. to represent a set of elements when deconstructing. Like so:

Yea, but I think it's a slice pattern in this case. .. is also used in range expressions. 
Not sure if slice operator / range operator Symbol needed the reused for spread expression. 
That's just my opinion / observation that is biased based on my experiences, so yea. I guess there wasn't solution that would content everyone.

> This is covered in the design doc :)
I could find this one! [1] Not sure if that's the document you were referring to. 

[1] https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-09.md#ambiguity-of--in-collection-expressions

**Reply by CyrusNajmabadi** [reply]
> Yea, but I think it's a slice pattern in this case

Right.  ANd we view 'spread' as the inverse of slicing.  You `slice` to pull out a subcollection, and you `spread` to put a subcollection into a larger one.

This is exactly the construction/deconstruction parity we're talking about.  

> Not sure if that's the document you were referring to.

The language is here: https://github.com/dotnet/csharplang/issues/5354

> An inclusive solution is needed for C#. It should meet the vast majority of casse for customers in terms of the collection-like types and values they already have. It should also feel natural in the language and mirror the work done in pattern matching.

> This leads to a natural conclusion that the syntax should be like [e1, e2, e3, e-etc] or [e1, ..c2, e2], which correspond to the pattern equivalents of [p1, p2, p3, p-etc] and [p1, ..p2, p3].
